homeassistant/helpers/integration_platform.py

```diff
@@ -36,75 +36,127 @@ class IntegrationPlatform:
-            component_name,
-            platform_name,
+def _async_integration_platform_component_loaded(
+    hass: HomeAssistant,
+    integration_platforms: list[IntegrationPlatform],
+    event: EventType[EventComponentLoaded],
+) -> None:
+    """Process integration platforms for a component."""
+    component_name = event.data[ATTR_COMPONENT]
+    if "." in component_name:
+        return
+
+    to_process: list[IntegrationPlatform] = []
+    for integration_platform in integration_platforms:
+        if component_name not in integration_platform.seen_components:
+            to_process.append(integration_platform)
+        integration_platform.seen_components.add(component_name)
+
+    if to_process:
+        hass.async_create_task(
+            _async_process_integration_platforms_for_component(
+                hass, component_name, to_process
+            ),
+            eager_start=True,
         )
-        return None
 
-    #
-    # Loading the platform may do quite a bit of blocking I/O
-    # and CPU work. (https://github.com/python/cpython/issues/92041)
-    #
-    # We don't want to block the event loop for too
-    # long so we check if the platform exists with `platform_exists`
-    # before trying to load it. `platform_exists` will do two
-    # `stat()` system calls which is far cheaper than calling
-    # `integration.get_platform`
-    #
-    if integration.platform_exists(platform_name) is False:
-        # If the platform cannot possibly exist, don't bother trying to load it
-        return None
 
-    try:
-        return integration.get_platform(platform_name)
-    except ImportError as err:
-        if f"{component_name}.{platform_name}" not in str(err):
-            _LOGGER.exception(
-                "Unexpected error importing %s/%s.py",
-                component_name,
-                platform_name,
-            )
+def _filter_possible_platforms(
+    integration: Integration,
+    integration_platforms: list[IntegrationPlatform],
+) -> list[IntegrationPlatform]:
+    """Filter out platforms that have already been processed.
 
-    return None
+    This function is executed in an executor.
+    """
+    return [
+        integration_platform
+        for integration_platform in integration_platforms
+        if integration.platform_exists(integration_platform.platform_name)
+    ]
 
 
-@callback
-def _async_process_integration_platforms_for_component(
+async def _async_process_integration_platforms_for_component(
     hass: HomeAssistant,
+    component_name: str,
     integration_platforms: list[IntegrationPlatform],
-    event: EventType[EventComponentLoaded],
 ) -> None:
     """Process integration platforms for a component."""
-    component_name = event.data[ATTR_COMPONENT]
-    if "." in component_name:
+    integration = await async_get_integration(hass, component_name)
+    # First filter out platforms that the integration already
+    # knows are missing
```

Comment:
We could make a get cached platform as well so we avoid the task it's already there.  Config could use it as well to offer a strong guarantee of no blocking io there